src/state/apis/nft/nftApi.ts

```diff
@@ -58,6 +63,126 @@ export const initialState: NftState = {
+      const nftItem: NftItem = {
+        ...nftItemWithoutId,
+        collectionId: item.collection.assetId,
+      }
+      acc[nftAssetId] = nftItem
+
+      return acc
+    }, {})
+
+    const collectionsById = resultsData.reduce<NftState['collections']['byId']>((acc, item) => {
+      if (!item.collection.assetId) return acc
+
+      const collectionAssetId = item.collection.assetId
+      if (!collectionAssetId) return acc
+
+      acc[collectionAssetId] = item.collection
+
+      return acc
+    }, {})
+
+    dispatch(
+      nft.actions.upsertCollections({
+        byId: collectionsById,
+        ids: Object.keys(collectionsById),
+      }),
+    )
+    dispatch(nft.actions.upsertNfts({ byId: nftsById, ids: Object.keys(nftsById) }))
+
+    const portfolioNfts = portfolioAssetIds.filter(isNft)
+    const missingPortfolioNfts = Object.values(nftsById).filter(
+      nft => !portfolioNfts.includes(nft?.assetId ?? ''),
+    )
+    const assetsToUpsert = missingPortfolioNfts.reduce<AssetsState>(
+      (acc, nft) => {
+        if (!nft) return acc
+
+        acc.byId[nft.assetId] = makeAsset({
+          assetId: nft.assetId,
+          id: nft.id,
+          symbol: nft.symbol ?? 'N/A',
+          name: nft.name,
+          precision: 0,
+          icon: nft.medias[0]?.originalUrl,
+        })
+
+        acc.ids.push(nft.assetId)
+
+        return acc
+      },
+      { byId: {}, ids: [] },
+    )
+
+    const portfolioDataToUpsert: Portfolio = {
+      accounts: {
+        byId: {},
+        ids: [],
+      },
+      accountBalances: {
+        byId: {},
+        ids: [],
+      },
+      accountMetadata: {
+        byId: {},
+        ids: [],
+      },
+      wallet: {
+        byId: {},
+        ids: [],
+      },
+    }
+
+    missingPortfolioNfts.forEach(nft => {
+      if (!nft) return
+
+      const accountId = nft.ownerAccountId
+
+      if (!accountId) return
+
+      if (!portfolioDataToUpsert.accounts.byId[accountId]) {
+        portfolioDataToUpsert.accounts.byId[accountId] = {
+          assetIds: [nft.assetId],
+        }
+        portfolioDataToUpsert.accounts.ids.push(accountId)
+      } else {
+        portfolioDataToUpsert.accounts.byId[accountId].assetIds.push(nft.assetId)
+      }
+
+      const balanceData = {
+        // i.e 1 for ERC-721s / 0, 1, or more for ERC-1155s
+        [nft.assetId]: nft.balance === undefined ? '1' : nft.balance,
+      }
+      if (!portfolioDataToUpsert.accountBalances.byId[accountId]) {
+        portfolioDataToUpsert.accountBalances.byId[accountId] = balanceData
+        portfolioDataToUpsert.accountBalances.ids.push(accountId)
+      } else {
+        portfolioDataToUpsert.accountBalances.byId[accountId] = Object.assign(
+          portfolioDataToUpsert.accountBalances.byId[accountId],
+          balanceData,
+        )
+      }
+    })
+
+    dispatch(assetsSlice.actions.upsertAssets(assetsToUpsert))
+    dispatch(portfolio.actions.upsertPortfolio(portfolioDataToUpsert))
```

Comment:
little bit of cleanup if you want.
- unless empty string `assetIds` are a thing (shouldn't be or they mean nothing to us), remove checks around this
- use a regular record instead of the partial to avoid `!nftItem` checks when we know we have them
- use `initialPortfolioState` object to reduce lines
- note need some import updates for this suggestion
```suggestion
    const nftsById = resultsData.reduce<Record<AssetId, NftItem>>((acc, item) => {
      const { collection, ...nftItemWithoutCollection } = item
      const nftItem: NftItem = {
        ...nftItemWithoutCollection,
        collectionId: item.collection.assetId,
      }
      acc[item.assetId] = nftItem
      return acc
    }, {})

    dispatch(nft.actions.upsertNfts({ byId: nftsById, ids: Object.keys(nftsById) }))

    const collectionsById = resultsData.reduce<NftState['collections']['byId']>((acc, item) => {
      if (!item.collection.assetId) return acc
      acc[item.collection.assetId] = item.collection
      return acc
    }, {})

    dispatch(
      nft.actions.upsertCollections({ byId: collectionsById, ids: Object.keys(collectionsById) }),
    )

    const portfolioNfts = portfolioAssetIds.filter(isNft)
    const missingPortfolioNfts = Object.values(nftsById).filter(
      nft => !portfolioNfts.includes(nft.assetId),
    )

    const assetsToUpsert = missingPortfolioNfts.reduce<AssetsState>(
      (acc, nft) => {
        acc.byId[nft.assetId] = makeAsset({
          assetId: nft.assetId,
          id: nft.id,
          symbol: nft.symbol ?? 'N/A',
          name: nft.name,
          precision: 0,
          icon: nft.medias[0]?.originalUrl,
        })
        acc.ids.push(nft.assetId)
        return acc
      },
      { byId: {}, ids: [] },
    )

    const portfolio = cloneDeep<Portfolio>(initialPortfolioState)

    missingPortfolioNfts.forEach(nft => {
      const accountId = nft.ownerAccountId

      if (!portfolio.accounts.byId[accountId]) {
        portfolio.accounts.byId[accountId] = { assetIds: [nft.assetId] }
        portfolio.accounts.ids.push(accountId)
      } else {
        portfolio.accounts.byId[accountId].assetIds.push(nft.assetId)
      }

      const balanceData = {
        // i.e 1 for ERC-721s / 0, 1, or more for ERC-1155s
        [nft.assetId]: nft.balance === undefined ? '1' : nft.balance,
      }
      if (!portfolio.accountBalances.byId[accountId]) {
        portfolio.accountBalances.byId[accountId] = balanceData
        portfolio.accountBalances.ids.push(accountId)
      } else {
        portfolio.accountBalances.byId[accountId] = Object.assign(
          portfolio.accountBalances.byId[accountId],
          balanceData,
        )
      }
    })

    dispatch(assetsSlice.actions.upsertAssets(assetsToUpsert))
    dispatch(portfolioSlice.actions.upsertPortfolio(portfolio))
```

Reply:
can push change to your branch if you would like